backend/matchAlgorithm.js

```diff
@@ -1,40 +1,143 @@
-import prisma from './prisma.js';
+import { PrismaClient } from '@prisma/client';
+import { fetchCosineSimilarity } from './TextSimilarity.js';
 
+const prisma = new PrismaClient();
 
-async function galeShapley() {
+export async function galeShapley(loggedInUserId, userRole) {
   const adopters = await prisma.adopter.findMany();
   const adoptees = await prisma.adoptee.findMany();
+  const rawAdopterPreferences = await prisma.adopterPreference.findMany();
+  const rawAdopteePreferences = await prisma.adopteePreference.findMany();
 
-  const adopterPreferences = await prisma.adopterPreferences.findMany();
-  const adopteePreferences = await prisma.adopteePreferences.findMany();
-
-  const matches = {};
+  const adopterPreferences = {};
+  const adopteePreferences = {};
 
   for (const adopter of adopters) {
-    const preferences = adopterPreferences.filter((preference) => preference.adopterId === adopter.id);
-    for (const preference of preferences) {
-      const adoptee = adoptees.find((adoptee) => adoptee.id === preference.adopteeId);
-      if (adoptee && !matches[adopter.id]) {
-        matches[adopter.id] = {
-          firstName:adoptee.firstName,
-          lastName:adoptee.lastName,
-        };
-        break;
-      }
-    }
+    const adopterId = adopter.id;
+    adopterPreferences[adopterId] = rawAdopterPreferences
+      .filter(preference => preference.adopterId === adopterId)
+      .sort((a, b) => b.rank - a.rank)
+      .map(preference => preference.adopteeId);
   }
 
   for (const adoptee of adoptees) {
-    const preferences = adopteePreferences.filter((preference) => preference.adopteeId === adoptee.id);
-    for (const preference of preferences) {
-      const adopter = adopters.find((adopter) => adopter.id === preference.adopterId);
-      if (adopter && !matches[adopter.id]) {
-        matches[adopter.id] = adoptee;
-        break;
+    const adopteeId = adoptee.id;
+    adopteePreferences[adopteeId] = rawAdopteePreferences
+      .filter(preference => preference.adopteeId === adopteeId)
+      .sort((a, b) => b.rank - a.rank)
+      .map(preference => preference.adopterId);
+  }
```

Comment:
Each Adopter's ID maps to an array of adoptee IDs which is sorted by preference in a descending order and vice versa